File: ruby/lib/mable_bank/application/ports/readers.rb
# frozen_string_literal: true

module MableBank
  module Application
    module AccountBalanceReader
      def each_account
        raise NotImplementedError, "#{self.class} must implement #each_account"
      end
    end

    module TransferInstructionReader
      def each_instruction
        raise NotImplementedError, "#{self.class} must implement #each_instruction"
      end
    end
  end
end

File: nodejs-typescript/src/application/ports/readers.ts
import type { Account } from '../../domain/Account.js';
import type { TransferInstruction } from '../../domain/TransferInstruction.js';

/** Port: stream opening balances into the domain. */
export interface AccountBalanceReader {
  eachAccount(callback: (account: Account) => void): void;
}

/** Port: stream transfer instructions for a processing day. */
export interface TransferInstructionReader {
  eachInstruction(callback: (instruction: TransferInstruction) => void): void;
}
